---
title: "Differentially Expressed Genes"
author: "Shannon Stiles and Parth Khatri"
date: "Due 12/12/2020"
output: word_document
---

```{r setup}
library(tidyverse)
library(dplyr)
library(Seurat)
library(stringr)
library(caret)
library(e1071)
```

# Notes and Workflow References
Processing: https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html
Differentially Expressed Genes: https://satijalab.org/seurat/v3.2/immune_alignment.html


# Check memory.size and expand memory.limit to 500 GB

```{r}

curr_memory = memory.size()
curr_memory

```
```{r}
# curr_memory = 272.90 MB
memory.limit(size = 2000 * curr_memory)

# memory.limit = 545780 MB

```
# Read in Seurat Object (QC and Cell Selection already performed)
```{r}

pbmc <- readRDS("./data/cs760project_COVID19scRNAseq_dataCohort2.rds")

```


# Normalize Data
"Global-scaling normalization method "LogNormalize" normalizes feature expression measurements for each cell by total expression, multiplies by a scale factor (10,000 by default), and log transforms the result"
```{r}
pbmc <- NormalizeData(pbmc)
```
# Identify highly variable features (feature selection)
"Subset of features that exhibit high cell-to-cell variation in the dataset"
To be used for PCA (linear dimensionality reduction)

```{r}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)
top10

```
## Scaling the Data

"Apply a linear transformation ('scaling') that is a standard pre-processing step prior to dimensional reduction techniques like PCA." 

The ScaleData function:
- Shifts the expression of each gene, so that the mean expression across - cells is 0
Scales the expression of each gene, so that the variance across cells is 1
- This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
- The results of this are stored in pbmc[["RNA"]]@scale.data

```{r}
pbmc <- ScaleData(pbmc, features = rownames(pbmc))
```
## Perform linear dimensional reduction (PCA)
Perform PCA on the scaled data. 
- by default only the previously determined variable features are used as input

```{r}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
```
# Examine and visualize PCA results
```{r}
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
```

```{r}
DimPlot(pbmc, reduction = "pca")
```

## Determine the 'dimensionality' of the dataset

"To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a 'metafeature' that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset."

‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each one (ElbowPlot function)."

```{r}
ElbowPlot(pbmc)
```

## Cluster the Cells

"Seurat v3 applies a graph-based clustering approach, building upon initial strategies in (Macosko et al). Importantly, the distance metric which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data [SNN-Cliq, Xu and Su, Bioinformatics, 2015] and CyTOF data [PhenoGraph, Levine et al., Cell, 2015]. Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'.

As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).

To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM [SLM, Blondel et al., Journal of Statistical Mechanics], to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters function implements this procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the Idents function."

```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:17)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```
## Run non-linear dimensional reduction (UMAP/tSNE)

"Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis."

```{r}
pbmc <- RunUMAP(pbmc, dims = 1:17)
```

```{r}
# individual clusters
DimPlot(pbmc, reduction = "umap")
```
```{r}
# individual clusters
DimPlot(pbmc, reduction = "umap", group.by = "covid")
```
```{r}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(pbmc, reduction = "umap", group.by = "group_per_sample")
```


```{r}

DimPlot(pbmc, reduction = "umap", split.by = "covid")
```

```{r}

DimPlot(pbmc, reduction = "umap", split.by = "group_per_sample")
```



# Identify differential expressed genes across conditions
"Use FindMarkers to find genes that are different between the all Covid and Control conditions"

```{r}

Idents(pbmc) <- pbmc_cluster@meta.data$'covid'
diff_genes <- FindMarkers(pbmc, ident.1 = "FALSE", ident.2 = "TRUE", verbose = FALSE)
head(diff_genes, n = 15)

```

# Identify differential expressed genes across conditions
"Use FindMarkers to find genes that are different between the Severe Covid and Control conditions"
```{r}
Idents(pbmc) <- pbmc_cluster@meta.data$'group_per_sample'
diff_genes_CS <- FindMarkers(pbmc, ident.1 = "control", ident.2 = "severe", verbose = FALSE)
head(diff_genes_CS, n = 15)

```

```{r}

# write differentially expressed genes to csv

path <- "./data/"

# write.csv(diff_genes, file.path(path, "Covid_Control_DE_genes.csv"),row.names = TRUE)

write.csv(diff_genes_CS, file.path(path, "DE_genes.csv"),row.names = TRUE)
```

# Exploratory
```{r}
# compare the two diff_genes sets

DF <- row.names(diff_genes)

DF_CS <- row.names(diff_genes_CS)

DE_genes <- DF_CS

# differences

differences <- setdiff(DF, DF_CS)

differences2 <- setdiff(DF_CS, DF)


# not in DF
only_DF <- unique(DF[! DF %in% DF_CS])

# not in DF_CS
only_DF_CS <- unique(DF_CS[! DF_CS %in% DF])

```